### PR TITLE
Kernel/aarch64: Stop hard-coding memory addresses (with the help of the devicetree)

### PR DESCRIPTION
**Pull request by spholz**
This is a first step toward removing all hard-coded addresses in the aarch64 kernel (and also making it able to boot on systems other than the Pi 3).

@Hendiadyoin1 I hope you are you are fine with me keeping you as the commit author of some of these commits which I took from #23703, even if I changed some of them quite a bit.

### CODE REVIEW

File: Kernel/Memory/MemoryManager.cpp [hunk 1]
```diff
@@ -518,31 +520,27 @@ UNMAP_AFTER_INIT void MemoryManager::parse_memory_map_fdt(MemoryManager::GlobalD
...
-                            state.size = data_as_int;
-                            data = data.slice(sizeof(u32));
-                        } else {
-                            BigEndian<u64> data_as_int;
-                            __builtin_memcpy(&data_as_int, data.data(), sizeof(u64));
-                            state.size = data_as_int;
-                            data = data.slice(sizeof(u64));
+                        FixedMemoryStream reg_stream { data };
+
+                        while (!reg_stream.is_eof()) {
```
Comment (Hendiadyoin1) [inline Kernel/Memory/MemoryManager.cpp:525]: Can we pre-allocate the state.reg here, as the eventual size is quite easy to calculate

File: Meta/run.py [hunk 2]
```diff
@@ -689,7 +689,7 @@ def set_up_network_hardware(config: Configuration):
 
 def set_up_kernel(config: Configuration):
     if config.architecture == Arch.Aarch64:
-        config.kernel_and_initrd_arguments = ["-kernel", "Kernel/Kernel"]
+        config.kernel_and_initrd_arguments = ["-kernel", "Kernel/kernel8.img"]
```
Comment (Hendiadyoin1) [inline Meta/run.py:694]: This change seems unrelated at this point

Comment (spholz) [reply Meta/run.py:694]: No, we need to switch to kernel8.img in this commit. I added a reason in the commit message.

File: Kernel/Memory/MemoryManager.cpp [hunk 3]
```diff
@@ -462,25 +464,25 @@ UNMAP_AFTER_INIT void MemoryManager::parse_memory_map_fdt(MemoryManager::GlobalD
                     state.state = State::Root;
                     break;
                 case State::InMemory:
-                    VERIFY(state.start.has_value() && state.size.has_value());
-                    dbgln("MM: Memory Range {}: address: {} size {:#x}", node_name, PhysicalAddress { state.start.value() }, state.size.value());
-                    global_data.physical_memory_ranges.append(PhysicalMemoryRange { PhysicalMemoryRangeType::Usable, PhysicalAddress { state.start.value() }, state.size.value() });
+                    for (auto const& reg_entry : state.reg) {
```
Comment (Hendiadyoin1) [inline Kernel/Memory/MemoryManager.cpp:466]: The preallocating might be possible here as well, as well as in the ReservedMemory one

**Review by Hendiadyoin1 — APPROVED** [review]
Just one Nit, otherwise LGTM

File: Kernel/Memory/MemoryManager.cpp [hunk 4]
```diff
@@ -462,26 +463,35 @@ UNMAP_AFTER_INIT void MemoryManager::parse_memory_map_fdt(MemoryManager::GlobalD
...
                     break;
                 case State::InMemory:
-                    VERIFY(state.start.has_value() && state.size.has_value());
-                    dbgln("MM: Memory Range {}: address: {} size {:#x}", node_name, PhysicalAddress { state.start.value() }, state.size.value());
-                    global_data.physical_memory_ranges.append(PhysicalMemoryRange { PhysicalMemoryRangeType::Usable, PhysicalAddress { state.start.value() }, state.size.value() });
+                    global_data.physical_memory_ranges.grow_capacity(global_data.physical_memory_ranges.size() + state.reg.size());
+
+                    for (auto const& reg_entry : state.reg) {
+                        dbgln("MM: Memory Range {}: address: {} size {:#x}", node_name, PhysicalAddress { reg_entry.start_addr }, reg_entry.size);
+                        global_data.physical_memory_ranges.append(PhysicalMemoryRange { PhysicalMemoryRangeType::Usable, PhysicalAddress { reg_entry.start_addr }, reg_entry.size });
```
Comment (Hendiadyoin1) [inline Kernel/Memory/MemoryManager.cpp:469]: This can be `unchecked_append` now, same in both loops below, afaics

File: Meta/run.py [hunk 5]
```diff
@@ -898,8 +905,7 @@ def configure_and_run():
 
     arguments = assemble_arguments(config)
 
-    build_directory = environ.get("SERENITY_BUILD", ".")
-    os.chdir(build_directory)
+    os.chdir(BUILD_DIRECTORY)
```
Comment (LucasChollet) [inline Meta/run.py:908]: If you have time to put that fix in its own commit, that would be nice :^)

**Review by Hendiadyoin1 — APPROVED** [review]

File: Meta/run.py (hunk 5 above)
Comment (nico) [reply Meta/run.py:908]: Doesn't look like this is done?

Comment (spholz) [reply Meta/run.py:908]: I moved the SERENITY_BUILD_DIR environment variable fix out of 730cfbd1f1b5fde5ca7c9d6508b556246d87e361 into a new commit aea8ff48219db53f206b6b85d680fe99eb838fe5.
Or was that review comment referring to something else?

Comment (nico) [reply Meta/run.py:908]: Oh sorry, I thought Lucas meant removing the `build_directory` local. Maybe I misunderstood what he meant, though!

Comment (LucasChollet) [reply Meta/run.py:908]: My bad it wasn't explicit, but fixing the name was my main concern. I don't mind where the other change ends up.

File: Kernel/Arch/init.cpp [hunk 6]
```diff
@@ -47,6 +47,7 @@
 #include <Kernel/FileSystem/VirtualFileSystem.h>
 #include <Kernel/Firmware/ACPI/Initialize.h>
 #include <Kernel/Firmware/ACPI/Parser.h>
+#include <Kernel/Firmware/DeviceTree/DeviceTree.h>
```
Comment (nico) [inline Kernel/Arch/init.cpp:50]: (re commit description: fine as-is, but in general it's nice to make code move commits contain nothing but a mechanical code move. tweaks and additions can be in later commits.)

**Review by nico — APPROVED** [review]
Very nice!
